blaze-client/src/main/scala/org/http4s/client/blaze/BlazeClient.scala

```diff
@@ -1,48 +1,45 @@
-package org.http4s.client.blaze
+package org.http4s
+package client
+package blaze
 
 import org.http4s.blaze.pipeline.Command
-import org.http4s.client.Client
-import org.http4s.{Request, Response}
 
 import scala.concurrent.duration.Duration
 import scalaz.concurrent.Task
-import scalaz.stream.Process.eval_
 import scalaz.{-\/, \/-}
 
 /** Blaze client implementation */
-final class BlazeClient(manager: ConnectionManager, idleTimeout: Duration, requestTimeout: Duration) extends Client {
-
-  /** Shutdown this client, closing any open connections and freeing resources */
-  override def shutdown(): Task[Unit] = manager.shutdown()
-
-  override def prepare(req: Request): Task[Response] = Task.suspend {
-    def tryClient(client: BlazeClientStage, freshClient: Boolean, flushPrelude: Boolean): Task[Response] = {
-      // Add the timeout stage to the pipeline
-      val ts = new ClientTimeoutStage(idleTimeout, requestTimeout, bits.ClientTickWheel)
-      client.spliceBefore(ts)
-      ts.initialize()
-
-      client.runRequest(req, flushPrelude).attempt.flatMap {
-        case \/-(r)    =>
-          val recycleProcess = eval_(Task.delay {
-            if (!client.isClosed()) {
-              ts.removeStage
-              manager.recycleClient(req, client)
+object BlazeClient {
+  def apply(manager: ConnectionManager, idleTimeout: Duration, requestTimeout: Duration): Client = {
+    Client(Service.lift { req =>
+      def tryClient(client: BlazeClientStage, freshClient: Boolean, flushPrelude: Boolean): Task[DisposableResponse] = {
+        // Add the timeout stage to the pipeline
+        val ts = new ClientTimeoutStage(idleTimeout, requestTimeout, bits.ClientTickWheel)
+        client.spliceBefore(ts)
+        ts.initialize()
+
+        client.runRequest(req, flushPrelude).attempt.flatMap {
+          case \/-(r)    =>
+            val dispose = Task.delay {
+              if (!client.isClosed()) {
+                ts.removeStage
+                manager.recycleClient(req, client)
+              }
             }
-          })
-          Task.now(r.copy(body = r.body.onComplete(recycleProcess)))
+            Task.now(DisposableResponse(r, dispose))
 
-        case -\/(Command.EOF) if !freshClient =>
-          manager.getClient(req, freshClient = true).flatMap(tryClient(_, true, flushPrelude))
+          case -\/(Command.EOF) if !freshClient =>
+            manager.getClient(req, freshClient = true).flatMap(tryClient(_, true, flushPrelude))
 
-        case -\/(e) =>
-          if (!client.isClosed()) client.shutdown()
-          Task.fail(e)
+          case -\/(e) =>
+            if (!client.isClosed()) client.shutdown()
+            Task.fail(e)
+        }
       }
-    }
 
-    val flushPrelude = !req.body.isHalt
-    manager.getClient(req, false).flatMap(tryClient(_, false, flushPrelude))
+      val flushPrelude = !req.body.isHalt
+      manager.getClient(req, false).flatMap(tryClient(_, false, flushPrelude))
+    }, manager.shutdown())
```

Comment:
I think the exposure is the exact same, I'm just pointing it out since we are refactoring and all.

I think the only alternative to an explicit shutdown is to let the GC handle it. In this case I don't think its the worst idea (in fact probably a good backup idea) but probably not good form for the only method of cleanup.